### PR TITLE
tests/kola/rpm-ostree: add test for kernel replacement

### PR DESCRIPTION
**Pull request by jmarrero**
closes https://github.com/coreos/rpm-ostree/issues/4205

### CODE REVIEW

**Comment by jmarrero** [comment]
This test passes on Fedora CoreOS now. I'll submit a follow up PR once: https://github.com/containers/skopeo/issues/1846 is fixed/resolved.

File: tests/kola/rpm-ostree/kernel-replace [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  1)
+    # Setup
+    # copy the OCI dir to containers-storage for a local build
+    skopeo copy $image containers-storage:localhost/fcos
+    rm "${image_dir}" -rf
+    td=$(mktemp -d)
+    cd ${td}
+    arch=$(arch)
+    version=$(rpm-ostree --version | grep Version)
+    kver="5.14.0-226.el9.${arch}"
```
Comment (cgwalters) [inline tests/kola/rpm-ostree/kernel-replace]: One thing we quickly learned after https://github.com/openshift/os/pull/1087 is that the kernel packages will get GC'd.  So...we will need to do one of:

- query dynamically
- if we're on rhel, replace with the latest centos kernel always  (but this doesn't help the test on centos)
- copy a fixed kernel package to some sort of lookaside place (tempting actually to stick it in a container image)

File: tests/kola/rpm-ostree/kernel-replace [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    derived=oci:$image_dir:derived
+    skopeo copy containers-storage:localhost/fcos-derived $derived
+    rpm-ostree --version
+    rpm-ostree rebase ostree-unverified-image:$derived
+    ostree container image list --repo=/ostree/repo
+    rm $image_dir -rf
+    /tmp/autopkgtest-reboot 2
+    ;;
+  2)
+    kver="5.14.0-226.el9.${arch}"
```
Comment (cgwalters) [inline tests/kola/rpm-ostree/kernel-replace]: Let's move this assignment to the top of the file so it's not duplicated.

File: tests/kola/rpm-ostree/kernel-replace [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    if uname -a | grep $kver; then
```
Comment (cgwalters) [inline tests/kola/rpm-ostree/kernel-replace]: Should be able to do something more like:
```
un=$(uname -r)
if test "$un" != "$kver"; then
   echo "Expected kernel $kver but found $un"
   exit 1
fi
```
or so?

File: tests/kola/rpm-ostree/kernel-replace (hunk 1 above)
Comment (jmarrero) [reply tests/kola/rpm-ostree/kernel-replace]: I can stick them on a container for sure. The only question is do we have anywhere to push containers like that? I can use my repo on quay.io or ghcr but we probably don't want that.

Comment (cgwalters) [reply tests/kola/rpm-ostree/kernel-replace]: We could create something like github.com/coreos/test-fixtures-container that has this and potentially other stuff for other tests?  It'd be pushed to `quay.io/coreos-assembler/test-fixtures` or so. Curious if anyone else on the team has opinons!

Comment (dustymabe) [reply tests/kola/rpm-ostree/kernel-replace]: What I don't fully understand here is why we're doing a kernel-replace of a centos kernel on FCOS. Technically it should work, but is that something we really want to test here? I wouldn't recommend any of our users do that.

My preference here would be that we pin on a Fedora kernel that is in our Fedora CoreOS Pool koji tag/distrepo. Those packages don't get GC until we manually do it once a Fedora cycle as part of the [rebase checklist](https://github.com/coreos/fedora-coreos-tracker/blob/main/.github/ISSUE_TEMPLATE/rebase.md). So we can just pin on one of those (for example [`kernel-6.0.18-300.fc37.x86_64.rpm`](https://kojipkgs.fedoraproject.org/repos-dist/coreos-pool/latest/x86_64/Packages/k/kernel-6.0.18-300.fc37.x86_64.rpm)) and then we update it once a cycle when we GC.

Comment (cgwalters) [reply tests/kola/rpm-ostree/kernel-replace]: Right, for Fedora we could also hardcode the "GA" kernel e.g. f37.  That's what we actually do in the rpm-ostree tests related to this.

But, we don't quite have that "GA" for centos stream AFAIK?

Comment (dustymabe) [reply tests/kola/rpm-ostree/kernel-replace]: Right now this test has `distros: fcos`, but I assume we want to run it on the others.

We should probably have a Fedora kernel that we use to replace on Fedora and a CentOS/RHEL kernel that we use on RHCOS/SCOS. For the Fedora kernel we can adopt the policy I mentioned above. For CentOS/RHEL we could just add a step for the user when they update the Fedora kernel to pick a new CentOS/RHEL kernel as well (from the location we're pulling them from).

I really don't think us managing our own copy of kernel files for this test (i.e. in a test fixtures container) as useful. There probably needs to be regular updating of the test to use a newer kernel for the test and the GC would force that to happen.

Comment (cgwalters) [reply tests/kola/rpm-ostree/kernel-replace]: > We should probably have a Fedora kernel that we use to replace on Fedora and a CentOS/RHEL kernel that we use on RHCOS/SCOS. For the Fedora kernel we can adopt the policy I mentioned above.

Yes, agree that's fine.

> For CentOS/RHEL we could just add a step for the user when they update the Fedora kernel to pick a new CentOS/RHEL kernel as well (from the location we're pulling them from).

That's the sticking point though, we don't have the same pool repo c9s side.  Arguably of course we should.  Right now looking at https://composes.stream.centos.org/development/ there's at least ~two months of history, so if we pick a new one now fixing the test every 2 months may not be so bad.

Or, we change this test to be dynamic and parse the composes.  May not be too hard, just ugly.

I think I'd say for now let's:

- Change this test to use the fedora kernel
- merge
- look at centos/rhel as a followup

?

Comment (jmarrero) [reply tests/kola/rpm-ostree/kernel-replace]: I just downloaded packages from: https://koji.fedoraproject.org/koji/buildinfo?buildID=1849252 we can probably use any of the koji builds and don't worry about updating this test for fedora for a year? maybe more. Would be nice to have a stable package to test against and not having to update a test every few months. @dustymabe why a using the coreos pool better?

Comment (jlebon) [reply tests/kola/rpm-ostree/kernel-replace]: Agree re. matching what rpm-ostree does: https://github.com/coreos/rpm-ostree/blob/c20e9834b35722679cf0fc02318966cfd7d71186/tests/vmcheck/test-override-kernel.sh#L36-L47.

> we can probably use any of the koji builds and don't worry about updating this test for fedora for a year? maybe more

I think that will work (at least until userspace adds requirements on newer kernels), but bigger picture I think for this test to stay realistic, we should be bumping it so it's tested against newer RPMs. Bumping it around rebase times seems reasonable to me.

Comment (dustymabe) [reply tests/kola/rpm-ostree/kernel-replace]: Yeah using the GA kernel seems like a good option.

Comment (dustymabe) [reply tests/kola/rpm-ostree/kernel-replace]: > I think I'd say for now let's:
> 
> * Change this test to use the fedora kernel
> * merge
> * look at centos/rhel as a followup

Works for me.

Comment (jmarrero) [reply tests/kola/rpm-ostree/kernel-replace]: PR updated.

**Review by cgwalters — APPROVED** [review]

**Comment by cgwalters** [comment]
```
[2023-01-14T17:24:46.234Z] --- PASS: ext.config.rpm-ostree.kernel-replace (434.87s)
```

**Pull request merged**